### PR TITLE
Add etag support for classpath resources

### LINKED ISSUE #2826: Add etag support for StaticFile.fromUrl where supported
`StaticFile.fromUrl` could support etags using the same mechanism as `StaticFile.fromFile` presuming the URL provided the following headers:

`Last-Modified`
`Content-Length`

While not every URL might provide those, most will.  This would add the ability to use etags when loading resources out of the classpath, which is a very common use case.

Discussed on Gitter here: https://gitter.im/http4s/http4s?at=5d682b5606e77851a313b3b0

Will try to see if it can get in 0.20 series.

### PR DESCRIPTION
**Pull request by matthughes**
Fixes #2826.  This is a replacement PR for #4879 that brings that PR up to date with main.

### CODE REVIEW

**Comment by matthughes** [comment]
Could use some assistance figuring out the scalafix error reported here: https://github.com/http4s/http4s/runs/7870927834?check_suite_focus=true

File: core/shared/src/main/scala/org/http4s/StaticFile.scala [hunk 1]
```diff
@@ -38,6 +37,8 @@ import org.typelevel.vault._
 import java.io._
 import java.net.URL
 
+import fs2.io._
```
Comment (hamnis) [inline core/shared/src/main/scala/org/http4s/StaticFile.scala:40]: group this with the other fs2.io imports then the scalafix warning should go away

Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/StaticFile.scala:40]: You can also run `scalafixAll` and it will fix it for you :)

**Comment by armanbilge** [comment]
Is it possible to target this at `series/0.23`? Then it can land in an 0.23.x release, currently it's slated for v1.0.0-M36.

File: core/shared/src/main/scala/org/http4s/StaticFile.scala (hunk 1 above)
Comment (matthughes) [reply core/shared/src/main/scala/org/http4s/StaticFile.scala:40]: I thought lintQuick would do that no?

**Comment by matthughes** [comment]
> Is it possible to target this at `series/0.23`? Then it can land in an 0.23.x release, currently it's slated for v1.0.0-M36.

@armanbilge I created a new PR at #6619 that targets 0.23.x.  I couldn't figure out how to do that while still giving @qwbarch credit but the PR is passing all checks now.  If that gets merged, should I then apply those changes to main in a new PR or is that something the team does?

**Comment by armanbilge** [comment]
> I couldn't figure out how to do that while still giving @qwbarch credit but the PR is passing all checks now.

It looks to me like qwbarch's work is already based on series/0.23:

https://github.com/http4s/http4s/compare/series/0.23...qwbarch:http4s:main

So instead of merging in `main` like you did in https://github.com/http4s/http4s/pull/6617/commits/9a4e1f17e714a6832d3d163d8570251d61d3cf4a, you would merge in `series/0.23`. I know it's annoying, but would you mind giving that another shot? 🙏

> If that gets merged, should I then apply those changes to main in a new PR or is that something the team does?

Yes, we periodically merge `series/0.23` into `main`, so another PR is not necessary :)

**Comment by matthughes** [comment]
@armanbilge Sorry I'm not following.  qwbarch's PR is based off his main branch.  If I change this PR to target series/0.23 I get a ton of merge conflicts.  Can we not use #6619?

**Comment by armanbilge** [comment]
> If I change this PR to target series/0.23 I get a ton of merge conflicts that have nothing to do with these changes.

@matthughes Yes, that's because you merged `main` into this PR in your commit https://github.com/http4s/http4s/pull/6617/commits/9a4e1f17e714a6832d3d163d8570251d61d3cf4a, so it is now far ahead of `series/0.23` :)

However, the original PR ends at commit https://github.com/http4s/http4s/commit/9a2333866f3296fdc86de48eafc0f0bfb987893e which is not ahead of `series/0.23`.

> Can we not use https://github.com/http4s/http4s/pull/6619?

Sorry to be a pain. I try to preserve original commit history and authorship when it's easy (i.e. doesn't require rebasing).

**Pull request closed without merging**